### PR TITLE
Merge in updates from vlasky fork

### PR DESCRIPTION
**Pull request by jswitzer**
Upgrade to the official MongoDB Go driver, a newer version of Go, and ignore the config DB on newer MongoDB versions.

### CODE REVIEW

File: lib/oplog/tail.go [hunk 1]
```diff
@@ -143,56 +172,137 @@ func (tailer *Tailer) tailOnce(out chan<- *redispub.Publication, stop <-chan boo
...
+		for {
+			gotResult, didTimeout, didLosePosition, err := readNextFromCursor(query)
 
-			if ts != nil {
-				lastTimestamp = *ts
-			}
+			if gotResult {
+				decodeErr := query.Decode(&rawData)
+				if decodeErr != nil {
+					log.Log.Errorw("Error decoding oplog entry", "error", decodeErr)
```
Comment (jgdef-tulip) [inline lib/oplog/tail.go:181]: do you have to return or otherwise handle the error here?

File: lib/oplog/tail.go [hunk 2]
```diff
@@ -143,56 +172,137 @@ func (tailer *Tailer) tailOnce(out chan<- *redispub.Publication, stop <-chan boo
 
 		var rawData bson.Raw
 
-		for iter.Next(&rawData) {
-			ts, pubs := tailer.unmarshalEntry(rawData)
+		for {
+			gotResult, didTimeout, didLosePosition, err := readNextFromCursor(query)
```
Comment (jgdef-tulip) [inline lib/oplog/tail.go:176]: i think any more than 4 values returned from a function, time to invent a data type to hold the result

File: lib/oplog/tail.go [hunk 3]
```diff
@@ -143,56 +172,137 @@ func (tailer *Tailer) tailOnce(out chan<- *redispub.Publication, stop <-chan boo
 
 		var rawData bson.Raw
```
Comment (jgdef-tulip) [inline lib/oplog/tail.go:173]: bring down to after line 178?

File: lib/oplog/tail.go [hunk 4]
```diff
@@ -143,56 +172,137 @@ func (tailer *Tailer) tailOnce(out chan<- *redispub.Publication, stop <-chan boo
...
+				// left off.
+				query, queryErr = issueOplogFindQuery(oplogCollection, lastTimestamp)
+
+				if queryErr != nil {
+					log.Log.Errorw("Error issuing tail query", "error", queryErr)
+					return
+				}
+
+				break
+			} else if err != nil {
```
Comment (jgdef-tulip) [inline lib/oplog/tail.go:220]: It's up to you, but rather than a longer if/else if ..../else block, something like the `else if err != nil` check could go right after line 176, and handling errors like that in an early-return block seems more go-ish

File: lib/oplog/tail.go [hunk 5]
```diff
@@ -143,56 +172,137 @@ func (tailer *Tailer) tailOnce(out chan<- *redispub.Publication, stop <-chan boo
...
+		time.Sleep(100 * time.Millisecond)
+		didTimeout = ctx.Err() != nil
+
+		// check if the error is a position-lost error. These errors are best handled
+		// by just re-issueing the query on the same connection; no need to surface
+		// a major error and re-connect to mongo
+		// From: https://github.com/rwynn/gtm/blob/e02a1f9c1b79eb5f14ed26c86a23b920589d84c9/gtm.go#L547
+		var serverErr mongo.ServerError
+		if errors.As(err, &serverErr) {
+			// 136  : cursor capped position lost
```
Comment (jgdef-tulip) [inline lib/oplog/tail.go:259]: OK... this seems version-specific...do these error codes exist since, say, mongodb 3.6? 4.0? And if 5.0 adds  more "lost position" error codes, we could miss that. But I guess in that case, we would just raise a generic error, and not optimize the lost position error.

File: lib/redispub/mongoTimestampEncoding.go [hunk 6]
```diff
@@ -4,32 +4,29 @@ import (
...
+func encodeMongoTimestamp(ts primitive.Timestamp) string {
+	return strconv.FormatUint(uint64(ts.T)<<32|uint64(ts.I), 10)
 }
 
-// Converts a string (in base-10) into a bson.MongoTimestamp
-func decodeMongoTimestamp(ts string) (bson.MongoTimestamp, error) {
-	i, err := strconv.ParseInt(ts, 10, 64)
+// Converts a string (in base-10) into a primitive.Timestamp
+func decodeMongoTimestamp(ts string) (primitive.Timestamp, error) {
+	i, err := strconv.ParseUint(ts, 10, 64)
```
Comment (jgdef-tulip) [inline lib/redispub/mongoTimestampEncoding.go:20]: Edge cases to test this?

File: blackbox-tests/docker-compose.yml [hunk 7]
```diff
@@ -33,14 +34,15 @@ services:
     volumes:
       - mongo_data:/data/db
   redis:
-    image: redis:3.2.4
+    image: redis:6.2.5
```
[the remark this answers is not in the document]
Comment (jswitzer) [reply blackbox-tests/docker-compose.yml:37]: Like the --platform =linux/amd64, this was so I could run the blackbox tests on my dev machine (M1 laptop). 

How do you feel about upgrading the version of redis we test against across the board? 

I think we've got runtime on it in dev at least:
Hori uses a somewhat modern redis from brew (depending on what users have installed, 6.2.6 on my machine), and runFactory uses bitnami/redis [12.1.1](https://github.com/bitnami/charts/blob/c1bfc0e28bf8a770855f3626b0386af8ae99479b/bitnami/redis/Chart.yaml), which encloses redis 6.0.9.

File: blackbox-tests/docker-compose.yml [hunk 8]
```diff
@@ -14,8 +14,9 @@ services:
       - sh
       - -c
       - |
-        mongo "$$OTR_MONGO_URL" --eval 'rs.initiate({ _id: "myapp", members: [{ _id: 0, host: "mongo:27017"}] })'
         update-ca-certificates
+        openssl s_client -connect redis_tls:6380
```
[the remark this answers is not in the document]
Comment (jswitzer) [reply blackbox-tests/docker-compose.yml:18]: I added openssl as a dependency in blackbox-tests/Dockerfile.oplogtoredis-blackbox. This s_client call is to debug the tls connection. I'm open to downgrading this to a comment and removing the openssl dependency, since this adds a bunch of logging, but if this test breaks again it's probably the first thing I'm going to check.

File: testapp/Dockerfile [hunk 9]
```diff
@@ -9,14 +20,14 @@ RUN meteor npm install && \
...
 # Install mongoDB client
-# The keys can get stale; to force the key import to re-run increment the "cache-bust" echo
-RUN echo "cache-bust 20190328" && apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 2930ADAE8CAF5059EE73BB4B58712A2291FA4AD5
-RUN echo "deb http://repo.mongodb.org/apt/debian jessie/mongodb-org/3.6 main" > /etc/apt/sources.list.d/mongodb-org-3.6.list
+RUN apt-get update && apt-get install -y gnupg wget
+RUN wget -qO - https://www.mongodb.org/static/pgp/server-4.4.asc | apt-key add -
+RUN echo "deb http://repo.mongodb.org/apt/debian buster/mongodb-org/4.4 main" > /etc/apt/sources.list.d/mongodb-org-4.4.list
 
-RUN apt-get update && apt-get install -y mongodb-org-shell=3.6.11 netcat && rm -rf /var/lib/apt/lists/*
+RUN apt-get update && apt-get install -y mongodb-org-shell=4.4.8 netcat && rm -rf /var/lib/apt/lists/*
```
[the remark this answers is not in the document]
Comment (jswitzer) [reply testapp/Dockerfile:30]: Looks like this comes from here, and it seems like it might not be necessary with the official debian images:

https://docs.docker.com/develop/develop-images/dockerfile_best-practices/

```
In addition, when you clean up the apt cache by removing /var/lib/apt/lists it reduces the image size, since the apt cache is not stored in a layer. Since the RUN statement starts with apt-get update, the package cache is always refreshed prior to apt-get install.

Official Debian and Ubuntu images [automatically run apt-get clean](https://github.com/moby/moby/blob/03e2923e42446dbb830c654d0eec323a0b4ef02a/contrib/mkimage/debootstrap#L82-L105), so explicit invocation is not required.
```

File: lib/oplog/tail.go [hunk 10]
```diff
@@ -113,26 +118,50 @@ func (tailer *Tailer) Tail(out chan<- *redispub.Publication, stop <-chan bool) {
...
+		findOneOpts := &options.FindOneOptions{}
+		findOneOpts.SetSort(bson.M{"$natural": -1})
+
+		queryContext, queryContextCancel := context.WithTimeout(context.Background(), config.MongoQueryTimeout())
+		defer queryContextCancel()
+
+		result := oplogCollection.FindOne(queryContext, bson.M{}, findOneOpts)
+
+		if result.Err() != nil {
+			return entry.Timestamp, result.Err()
```
[the remark this answers is not in the document]
Comment (jswitzer) [reply lib/oplog/tail.go:142]: Agreed this is too implicit - we're returning a timestamp of zero from this function in the error case, and we should make the caller handle the nil case explicitly. For example, we wouldn't want a timestamp acting as a cursor to be reset to zero by a parsing error (or if we want that behavior it should be made explicit).

File: blackbox-tests/docker-compose.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (jswitzer) [reply blackbox-tests/docker-compose.yml]: I cleaned this up - since we're only using redis 6, I removed the ssl proxy container from the docker compose test

File: blackbox-tests/Dockerfile.oplogtoredis-blackbox [hunk 11]
```diff
@@ -1,9 +1,14 @@
-FROM golang:1.12.9-alpine3.9
+FROM --platform=linux/amd64 golang:1.17.0-alpine3.14
```
[the remark this answers is not in the document]
Comment (jswitzer) [reply blackbox-tests/Dockerfile.oplogtoredis-blackbox]: Resolved - I added this to the other places that needed this to run on the M1. This is needed for debian-based images on which we later install mongodb.  Without the platform specifier you'll get an arm debian, and mongodb doesn't ship a debian arm repo.

File: blackbox-tests/docker-compose.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (jswitzer) [reply blackbox-tests/docker-compose.yml]: Resolved -  Standardized on redis 6.0, which has an arm image

File: Dockerfile.dev [hunk 12]
```diff
@@ -1,11 +1,16 @@
...
+FROM golang:1.17.0-alpine3.14
 
 ADD scripts/wait-for.sh /wait-for.sh
 
-RUN apk --update --no-cache add git openssh mongodb gcc cyrus-sasl cyrus-sasl-dev musl-dev && \
+RUN apk --update --no-cache add git openssh gcc cyrus-sasl cyrus-sasl-dev musl-dev && \
     mkdir -p /oplogtoredis && \
     go get github.com/pilu/fresh
 
+RUN echo 'http://dl-cdn.alpinelinux.org/alpine/v3.6/main' >> /etc/apk/repositories
```
[the remark this answers is not in the document]
Comment (jswitzer) [reply Dockerfile.dev]: (moving up from another thread) If we do care about the size of the images, we could run `rm -rf /var/lib/apt/lists/*` in more places

File: integration-tests/acceptance/transaction_test.go [hunk 13]
```diff
@@ -46,7 +44,7 @@ func TestTransaction(t *testing.T) {
 	// not easy to explicitly create a collection with this library:
 	// https://jira.mongodb.org/browse/GODRIVER-1147
```
[the remark this answers is not in the document]
Comment (jswitzer) [reply integration-tests/acceptance/transaction_test.go]: Resolved - Test updated to use explicit collection creation.

File: integration-tests/fault-injection/harness/mongo.go [hunk 14]
```diff
@@ -103,22 +103,29 @@ func (server *MongoServer) Start() {
 	}
 }
 
-// Client returns an mgo.Session configured to talk to the replica set
-func (server *MongoServer) Client() *mgo.Session {
-	dialInfo, err := mgo.ParseURL(server.Addr)
+// Client returns an mongo.Client configured to talk to the replica set
+func (server *MongoServer) Client() *mongo.Client {
+	cs, err := connstring.ParseAndValidate(server.Addr)
```
[the remark this answers is not in the document]
Comment (jswitzer) [reply integration-tests/fault-injection/harness/mongo.go:108]: I like this kind of flow strategy here, otherwise you'd have a lot of nested `{}`

File: integration-tests/fault-injection/harness/otr.go [hunk 15]
```diff
@@ -51,7 +52,11 @@ func StartOTRProcessWithEnv(mongoURL string, redisURL string, port int, extraEnv
 func (proc *OTRProcess) Start() {
 	log.Printf("Starting up oplogtoredis with HTTP on %d", proc.port)
 
-	proc.cmd = exec.Command("/bin/oplogtoredis") // #nosec
+	otrBin := os.Getenv("OTR_BIN")
```
[the remark this answers is not in the document]
Comment (jswitzer) [reply integration-tests/fault-injection/harness/otr.go:55]: I'll add a note in the fault-injection Dockerfile about this.

File: integration-tests/fault-injection/resume_test.go [hunk 16]
```diff
@@ -30,10 +31,14 @@ func TestResume(t *testing.T) {
 
 	// We insert a couple things into the oplog to make sure they don't
 	// get processed by oplogtoredis
-	testCollection := mongoClient.DB("").C("Test")
-	require.NoError(t, testCollection.Insert(bson.M{"_id": "id1"}))
-	require.NoError(t, testCollection.Insert(bson.M{"_id": "id2"}))
-	require.NoError(t, testCollection.Insert(bson.M{"_id": "id3"}))
+	testCollection := mongoClient.Database(mongo.DBName).Collection("Test")
+
+	_, err := testCollection.InsertOne(context.Background(), bson.M{"_id": "id1"})
```
[the remark this answers is not in the document]
Comment (jswitzer) [reply integration-tests/fault-injection/resume_test.go:36]: I like it as is, as the purpose of this testing code is imperative, the important act here is loading the data, not that the data is loaded. For example I could see us adding random sleeps or putting other things between the imports in the future.

File: integration-tests/meteor/array_test.go [hunk 17]
```diff
@@ -41,14 +43,15 @@ func TestArrayModification(t *testing.T) {
...
+	received := meteor1.VerifyReceive(t, harness.DDPMsgGroup{
 		harness.DDPResult("methodCallId", harness.DDPData{}),
 
 		expectedChange1,
 		expectedChange2,
-	}, harness.DDPMsgGroup{
+
 		harness.DDPUpdated([]string{"methodCallId"}),
 	})
+	received.VerifyUpdatedComesAfterAllChanges(t)
```
[the remark this answers is not in the document]
Comment (jswitzer) [reply integration-tests/meteor/array_test.go:54]: Yes, this appears to be an improvement on the original test. 

```
// Ensures that the given message group (which should be the set of messages
// that are sent in response to a Method) has no modification messages (added/changed/removed)
// coming *after* the "updated" message -- in DDP, the "updated" message indicates
// that all modifications resulting from the method have been sent
```

File: integration-tests/meteor/array_test.go [hunk 18]
```diff
@@ -13,6 +13,8 @@ func TestArrayModification(t *testing.T) {
 
 	require.NoError(t, meteor1.Send(harness.DDPMethod("insertCall", "arrayTest.initializeFixtures")))
 
+	meteor1.ClearReceiveBuffer()
```
[the remark this answers is not in the document]
Comment (jswitzer) [reply integration-tests/meteor/array_test.go:16]: This seems good - the test harness has a buffer that builds up changes, and then has a different method to check that the changes sent back from the meteor server match what we expect to have changed. 

The test follows this pattern: 
1. Establish DDP connection to meteor
2. Make sure that the system is 'quiescent' (AKA meteor isn't sending us active change updates)
3. Send some changes to the meteor server (or otherwise trigger changes in mongo)
4. Wait for the meteor server to send updates over DDP
5. Wait for the system to be quiescent again, then compare the updates to what we expect meteor would have told us.

File: integration-tests/meteor/harness/ddpConn.go [hunk 19]
```diff
@@ -187,3 +190,20 @@ func (conn *DDPConn) recvDDP() (*DDPMsg, error) {
...
+
+// Ensures that the given message group (which should be the set of messages
+// that are sent in response to a Method) has not modification messages (added/changed/removed)
+// coming *after* the "updated" message -- in DDP, the "updated" message indicates
+// that all modifications resulting from the method have been sent
+func (msgGroup DDPMsgGroup) VerifyUpdatedComesAfterAllChanges(t *testing.T) {
+	updatedReceived := false
+	for _, msg := range msgGroup {
+		if msg.DDPType == "updated" {
+			updatedReceived = true
```
[the remark this answers is not in the document]
Comment (jswitzer) [reply integration-tests/meteor/harness/ddpConn.go:202]: Yep, continue added. Also fixed typo

File: lib/oplog/tail.go (hunk 5 above)
Comment (jswitzer) [reply lib/oplog/tail.go:259]: I agree. This is just an optimization for errors we can handle by just re-creating the cursor.

File: lib/oplog/tail.go (hunk 2 above)
Comment (jswitzer) [reply lib/oplog/tail.go:176]: Yep, I agree - refactored this a bit and added a comment ..

File: lib/oplog/tail.go (hunk 1 above)
Comment (jswitzer) [reply lib/oplog/tail.go:181]: I think logging the error and continuing is the best we can do here, especially if we get a single bad-apple unparseable oplog entry I wouldn't want it to bring down the service and block subsequent updates.

File: lib/oplog/tail.go (hunk 3 above)
Comment (jswitzer) [reply lib/oplog/tail.go:173]: Yeah, I think so - I want to be clear that this isn't reused between loops..

File: lib/oplog/tail.go (hunk 1 above)
Comment (jswitzer) [reply lib/oplog/tail.go:181]: I added a 'continue' just to be explicit
